### PR TITLE
Upgrade CI

### PR DESCRIPTION
**Pull request by Luthaf**
This does multiple upgrades to the CI environement:

- ~Use M1 runners for macOS CI;~ EDIT: no longer part of this PR, will be done later
- Use rustc 1.75 to build the linux wheels 
- Use Python 3.12 instead of 3.11 as out highest supported version
- Use Torch 2.2 for the C++ tests

# Contributor (creator of pull-request) checklist

 - [ ] ~Tests updated (for new features and bugfixes)?~
 - [ ] ~Documentation updated (for new features)?~
 - [ ] ~Issue referenced (for PRs that solve an issue)?~

# Reviewer checklist

 - [ ] CHANGELOG updated with public API or any other important changes?

----
📚 Documentation preview 📚: https://metatensor--503.org.readthedocs.build/en/503/

### CODE REVIEW

File: python/scripts/rustc-manylinux2014_x86_64/Dockerfile [hunk 1]
```diff
@@ -5,7 +5,7 @@ RUN yum install git -y
 RUN git config --global --add safe.directory /code
 
 # Download rustup-init asn install
-RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --profile minimal --default-toolchain 1.65
+RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --profile minimal --default-toolchain 1.75
```
Comment (Luthaf) [inline python/scripts/rustc-manylinux2014_x86_64/Dockerfile:8]: the wheel build step went from 3min15s to 27s, nice!

**Comment by Luthaf** [comment]
> Use M1 runners for macOS CI;

One nice side effect for this is that we can now run PyTorch tests with non-CPU device on CI, using the MPS device. This should help finding device issues!

**Comment by PicoCentauri** [comment]
> EDIT: I might have spoke too soon, PyTorch fails to allocate MPS memory: [actions/runner-images#9254 (comment)](https://github.com/actions/runner-images/issues/9254#issuecomment-1936326374)

Unfortunate, So should we at least bump the versions of the Python interpreter as you already did in the CI?

**Comment by Luthaf** [comment]
yes, if we don't get an answer from github in the issue above, I'll keep only the other changes.

**Pull request merged**